codegen.cpp

```diff
@@ -227,22 +344,43 @@ void NodeNullStmt::codegen(){
 static void emit_prologue() {
     cout << "  stp x29, x30, [sp, -16]!" << endl;
     cout << "  mov x29, sp" << endl;
-    int stackSize = round16(var_map.size()*8);
+
+    // Allocate stack space for local variables
+    int stackSize = round16(var_map.size() * 8);
     cout << "  sub sp, sp, #" << stackSize << endl;
 }
 
 static void emit_epilogue() {
-    cout << ".L.return" << ":" << endl;
-    cout << "  add sp, sp, #" << round16(var_map.size()*8) << endl;
+    cout << ".L.return." << current_func_def_codegen->declarator << ":" << endl;
+    cout << "  add sp, sp, #" << round16(var_map.size() * 8) << endl;
     cout << "  ldp x29, x30, [sp], 16" << endl;
+    cout << "  mov " << return_reg << ", " << accum_reg << endl;;
     cout << "  ret" << endl;
 }
 
+void NodeFunctionDef::codegen() {
+    cout << ".global " << declarator << endl;
+    cout << declarator << ":" << endl;
+
+    emit_prologue();  // Emit prologue for this function
+    // Mapping arguments to local variables
+    for (size_t i = 0; i < params.size(); ++i) {
+        int offset = var_map[params[i]->varName]->offSet;  // Get stack offset for the parameter
+        if (i < 8) {
+            emit_store_to_stack("x" + to_string(i), offset);  // Save to stack
```

Comment:
so the arguments get passed in by the caller from x0 to x8 (or up to number of parameters), then you access function parameters by offsetting from the frame pointer. this works, but defeats the whole purpose of reserving x0-x8 in the first place, since you're not using them to access local variables anyway.